wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java

```diff
@@ -85,36 +98,36 @@ public WorkspaceRuntimes(MachineManager machineManager, EventService eventServic
         this.eventService = eventService;
         this.descriptors = new HashMap<>();
         this.startQueues = new HashMap<>();
-        this.rwLock = new ReentrantReadWriteLock();
+        this.addMachineEventSubscriber = new AddMachineEventSubscriber();
+        this.removeMachineEventSubscriber = new RemoveMachineEventSubscriber();
     }
 
     /**
      * Returns the runtime descriptor describing currently starting/running/stopping
      * workspace runtime.
      *
      * <p>Note that the {@link RuntimeDescriptor#getRuntime()} method
-     * returns {@link Optional} which describes just a snapshot copy of
-     * a real {@code WorkspaceRuntime} object, which means that any
-     * runtime copy modifications won't affect the real object and also
-     * it means that copy won't be affected with modifications applied
+     * returns a copy of a real {@code WorkspaceRuntime} object,
+     * which means that any runtime copy modifications won't affect the
+     * real object and also it means that copy won't be affected with modifications applied
      * to the real runtime workspace object state.
      *
      * @param workspaceId
      *         the id of the workspace to get its runtime
      * @return descriptor which describes current state of the workspace runtime
      * @throws NotFoundException
-     *         when workspace with given {@code workspaceId} doesn't have runtime
+     *         when workspace with given {@code workspaceId} is not running
      */
     public RuntimeDescriptor get(String workspaceId) throws NotFoundException {
-        rwLock.readLock().lock();
+        acquireReadLock(workspaceId);
```

Comment:
try with resources:
try (acquireReadLock(workspaceId) {


} // do not need to release the lock